### ISSUE TITLE
update mutation doesn't work 

### ISSUE THREAD

**Issue by masoud-moghini** [apollographql/apollo-client#4031]
**Intended outcome:**
updated list of recipes be rendered

**Actual outcome:**
the old list get rendered

**How to reproduce the issue:**
follow what comes [here](https://www.apollographql.com/docs/react/essentials/mutations.html#update) !!!!

**Versions**

hi I have the following client
```
import React ,{Component}from 'react'
import { Mutation } from "react-apollo";
import {ADD_RECIPE} from '../../mutations';
import Error from '../Error';
import {withRouter} from 'react-router-dom';
import {GET_ALL_RECIPIES} from '../../queries'

class AddRecipe extends Component {
  ...
  onSubmit(event,addRecipe){
    event.preventDefault();
    addRecipe().then(
      ({data})=>
        {
          this.props.history.push("/")
        }
      )
  }
  render (){
    const{name,category,description,instruction,username} = this.state;
    return(<div className="App">
      <h2>Add recipe</h2>
      <Mutation
             mutation={ADD_RECIPE} 
             variables={{name,category,description,instruction,username}} 
             update={(cache, {data:{addRecipe}}) => {
                  const {getAllRecipes} = cache.readQuery({ query: GET_ALL_RECIPIES });
                  console.log(cache)
                  //console.log('get all recipes',getAllRecipes)
                  //console.log('add recipe',addRecipe)
                  cache.writeQuery({
                    query:GET_ALL_RECIPIES,
                    data:{getAllRecipes:getAllRecipes.concat[addRecipe]}
                  })
                  console.log(cache)
              }} >
      {(addRecipe, {data,loading,error})=>
        
           (<form className="form" onSubmit={event=>this.onSubmit(event,addRecipe)}>
                  <input type="text" name="name" onChange={this.handleChange.bind(this)} placeholder="recipe name" value={name}/>
                  <select name="category" value="breakfast" id="" onChange={this.handleChange.bind(this)} value={category}>
                    <option value="breakfast">breakfast</option>
                    <option value="lunch">lunch</option>
                    <option value="dinner">dinner</option>
                    <option value="snack">snack</option>
                  </select>
                  <input type="text" name="description" onChange={this.handleChange.bind(this)} placeholder="recipe description" value={description}/>
                  <textarea name="instruction" id="instruction" cols="30" rows="10" onChange={this.handleChange.bind(this)} value={instruction}> </textarea>
                  <button type="submit" className="button-primary" disabled = {loading || this.validateForm()}>submit</button>
                  {error && <Error error={error}/>}
            </form>)
        
      }
      </Mutation>
    </div>)
  }
}
export default withRouter(AddRecipe)
```
and for get Recipe is:

```
import gql from 'graphql-tag';


//RECIPE QUERY
export const GET_ALL_RECIPIES =gql`
query {
    getAllRecipes{
        name
        _id
        category
        
    }
}
`

```

and get recipe item is : 

```
import gql  from "graphql-tag";

export const GET_RECIPE_ITEM =gql`
query($id:ID!){
        getRecipeItem(_id:$id){
    				
          _id
          name
          category
          description
          instruction
          createdDate
          likes
          username
		}
}`
```

**Comment by OurMajesty** [comment]
I think this is the same as #3992 
Try apollo-cache-inmemory 1.2.10

**Comment by benjamn** [comment]
Please try updating to the latest `apollo-cache-inmemory` and `apollo-client`:
```bash
npm install apollo-client@latest apollo-cache-inmemory@latest
```

**Comment by masoud-moghini** [comment]
> ```shell
> npm install apollo-client@latest apollo-cache-inmemory@latest
> ```

tried it and no result achieved.

**Comment by vu-bui** [comment]
same problem with `apollo-client` `2.4.3`.
reverting back to `2.4.2` works for me.

**Comment by bogdansoare** [comment]
having the same issue

**Comment by masoud-moghini** [comment]
have the same problem in optimistic ui about delete mutation.
real time page refreshing doesn't work for react-apollo and this is a great shortcoming.

**Comment by Akryum** [comment]
It seems sometimes adding an optimistic response fixes the issue: https://github.com/Akryum/vue-apollo/issues/436

**Comment by benjamn** [comment]
@Akryum Thanks for the reproduction! Interestingly, the problem disappears for me when I apply https://github.com/apollographql/apollo-client/pull/4069, though I don't have a complete theory as to why that makes a difference.

**Comment by benjamn** [comment]
@Akryum's reproduction is fixed by `apollo-client@2.4.5`, thanks to my PR #4069. Everyone else, please update your packages and try again. If the problem persists, a runnable CodeSandbox reproduction would really help us track it down.

**Comment by dsanders11** [comment]
@benjamn, can confirm that `apollo-client@2.4.5` fixes this issue for me.

However, I do notice that when deleting items, `writeQuery` isn't necessary to achieve the desired effect, probably because the result of `readQuery` is not a deep copy, and so modifying it (in this case, removing an item from an array) affects the store. This is with `vue-apollo` (and in turn `apollo-cache-inmemory`). Is this a known issue, or should I open a new issue? It seems that `readQuery` should logically return a deep copy for the in-memory cache, as otherwise behavior would be inconsistent with a different cache implementation.

**Comment by flytedan** [comment]
@Akryum thanks man! Yes the `apollo-client@2.4.5` fixes the problem!

**Comment by masoud-moghini** [comment]
I got its function confused with `refetchQuery` and if someone explains me what is the necessity of updating cache I would be grateful .

**Comment by benjamn** [comment]
@dsanders11 To clarify, modifying result objects returned by `readQuery` only affects the cached result objects, not the data in the store, which can only be modified by writing new results.

Modifying data and then writing it back to the store is safe, since it invalidates the previous result objects, so `readQuery` will return new objects next time. That's what the official docs are doing, unless I'm mistaken.

The problem with a deep copy is that it's expensive, both because the copy itself takes time, and because identical results are no longer `===` identical, which makes equality comparisons in your application (e.g. to determine whether to re-render a component) more expensive. Making a deep copy is something you (as the consumer of the cache) can choose to do, if you want to pay for that convenience / peace of mind. In practice, when you copy your results, you don't usually need a full deep copy, because you can get away with shallow-copying parts of the result and modifying the partial copy without affecting the original object. Since the cache doesn't know what parts of the data you're planning to change, it has no hope of making these decisions correctly. Only you understand the needs of your application. The job of the cache is to enable all use cases, without making assumptions about where you might value safety over performance. If the cache itself always made a deep copy whether or not you needed it, developers who don't want/need the copying behavior would be stuck with it anyway. In this area of the system, we've chosen to prioritize performance and developer choice over eliminating the possibility of misuse.

For those reasons, I think freezing the result objects is the only viable option here, though that would interfere with valid read-modify-write patterns, and would definitely require changing the documentation.

**Comment by ECIZEP** [comment]
@benjamn why are result objects returned by `readQuery` not  consistent with the data in the store? 
`readQuery` don't get data from the store, but return the cached result objects，so if i modify some items, there is no need to `writeQuery` to update the store，modifying the result objects is just enough。it seems that there are two cache layers，one is the result objects returned by `readQuery`, the other one is the store. This really makes me confused.
I also think the result objects should be immutable and frozen, or `readQuery` should return data from the store every time instead of return the result objects.

**Comment by charlie632** [comment]
Why is this issue still opened? It was resolved for me by installing the latests versions of apollo-client and apollo-cache-inmemory

``
npm install apollo-client@latest apollo-cache-inmemory@latest
``

**Comment by timbrandin** [comment]
I'm using these: 
```
apollo-cache-inmemory@1.3.11
apollo-client@2.4.7
```
=> and updates are not fired when running writeQuery.

**Comment by timbrandin** [comment]
I just switched back to ReduxCache, and that resolved my issue, not sure I can trust InMemoryCache right now due to this.

**Comment by scottie-schneider** [comment]
Same issue, using 
  "apollo-cache-inmemory": "^1.6.2",
  "apollo-client": "^2.6.2",

**Comment by leofuturer** [comment]
@masoud-moghini Thanks for your reproduction, but the codesandbox you provide does not work well now. It shows an error _Error sending the query 'items' TypeError: Failed to fetch_ , it seems that the server has something wrong.

**Comment by JamieKudla** [comment]
Was having the same issue with the UI not reflecting the updated cache. Following [the example in the docs](https://www.apollographql.com/docs/react/api/react-hoc/#optionsupdate), I could not get it working unless I created a new array for `data.todos`:

```
    options: {
      update: (proxy, { data: { createTodo } }) => {
        const data = proxy.readQuery({ query });
        // data.todos.push(createTodo);
        data.todos = [...data.todos, createTodo];
        proxy.writeQuery({ query, data });
      },
    },
```
```
"apollo-cache-inmemory": "^1.6.3",
"apollo-client": "^2.6.4",
```

**Comment by EdmundMai** [comment]
i had this problem too, then i inspected the record returned by the mutation and noticed that it was missing a field that the records in the cache had. i added that field to the mutation response and it worked

**Comment by finaxar-arthur** [comment]
@EdmundMai I thought I have the same problem. But I double checked my fields and created new array as @JamieKudla did.

Still didn't solve my issue.

**Comment by crisward** [comment]
Not sure if this helps, but I found it works if my collection didn't have id's returned. When it did have id's I had to write a fresh object.

```
apollo-client@2.6.4
@apollo/react-hooks@3.1.2
apollo-boost@0.4.4
apollo-cache-inmemory@1.6.3
```

eg

```javascript
// works with just re-creating array when no id's
export const GET_BOOKS = gql`{ 
  books { 
    title 
    author
  } 
}`

addBook({
  variables: {input:{ title: title.value, author: author.value }},
  update: (store, { data: { book } }) => {
    const data = store.readQuery({ query: GET_BOOKS }) // read books from cache
    data.books = [ ...data.books, book] // create new array with our book
    store.writeQuery({ query: GET_BOOKS, data}) // write back data
  },
})
```

```javascript
// with id, new data object needed
export const GET_BOOKS = gql`{ 
  books { 
    id
    title 
    author
  } 
}`

addBook({
  variables: {input:{ title: title.value, author: author.value }},
  update: (store, { data: { book } }) => {
    const data = store.readQuery({ query: GET_BOOKS }) // read books from cache
    const books = [ ...data.books, book] // create new array with our book
    store.writeQuery({ query: GET_BOOKS, data:{...data,books} }) // write back new object with our books
  },
})
```

**Comment by dokicro** [comment]
I have the same problem but what is the different first time I add a product it updates but second it stops working...

If I `console.log(store.readQuery)` it shows correct data just UI is not reflecting it...

So as a workaround because I have to push to production on Friday I created this:

```
const hasTransaction = this.bankAccount.transactions
              .find((transaction) => {
                return transaction.id === response.data.createBankTransaction.id;
              });

            if (!hasTransaction) {
              this.bankAccount.transactions.unshift(response.data.createBankTransaction);
            }
```

TLDR: If data is not added by write query I add it manually...

**Comment by thminggg** [comment]
Hi @crisward & all,
thanks for the comments up there. after your findings, is there any rule to follow to ensure the UI update accordingly?

like, always make new object or something?

the behaviour is so unpredictable, it's a nightmare...

**Comment by xhlsrj** [comment]
FYR,
`"@apollo/client": "^3.0.0-beta.49"`
When I use `update` like this, it does not work.
```
client.mutate({
// mutation,

update(cache){
  // const query;
  const data = cache.readQuery({ query });

  data.array.push({ key: value });

  cache.writeQuery({
    query,
    data,
  });
},
```
After I set a new object to data like this, it works.
```
update(cache){
  // ...
  cache.writeQuery({
    query,
    data: { array: [...data.array, { key: value }] },
  });
}
```

**Comment by thminggg** [comment]
Yup, always ensure the original cache object is immutable!
Assign new value by deep clone the object and assign new values will work!
